File: infrastructure/lambda/list-photos.ts
import {APIGatewayProxyEventV2, APIGatewayProxyResultV2} from 'aws-lambda';
import {S3} from 'aws-sdk';
import {HeadObjectRequest} from "aws-sdk/clients/s3";

// const bucketName = process.env.PHOTO_BUCKET_NAME!;
const bucketName = 'memerson-api-photos'
const s3 = new S3();

// TODO: Used gcd to caluclate aspect ratio
// TODO: return aspect ratio in addition to width and height
const gcd = (...arr: number[]): number => {
  const _gcd = (x: number, y: number): number => (!y ? x : gcd(y, x % y));
  return [...arr].reduce((a, b) => _gcd(a, b));
};

const getMetadata = async (object: S3.Object): Promise<{
  filename: string,
  url: string,
  width: number,
  height: number,
  aspect_ratio_width: number,
  aspect_ratio_height: number
}> => {
  const url = `https://${bucketName}.s3.amazonaws.com/${object.Key}`;

  const regexp = /(?<width>[0-9]+)x(?<height>[0-9]+)/;
  const match = object.Key?.match(regexp)?.groups;

  // const params: HeadObjectRequest = {
  //   Bucket: bucketName, Key: object.Key!
  // }

  // const {Metadata: metadata} = s3.headObject(params).promise();
  // console.log(`Metadata: ${JSON.stringify(metadata)}`);
  // const width: number = +(metadata!['width'] || match?.width || '3');
  // const height: number = +(metadata!['height'] || match?.width || '2');


  const width = +(match?.width || '3');
  const height = +(match?.height || '2');
  console.log(`Width: ${width} / Height: ${height}.`);
  const i = gcd(width, height)
  const aspect_ratio_width = width / i
  const aspect_ratio_height = height / i

  return {
      filename: object.Key!,
      url,
      width,
      height,
      aspect_ratio_width,
      aspect_ratio_height
  }
}

export async function handler(event: APIGatewayProxyEventV2): Promise<APIGatewayProxyResultV2> {
  console.log('event', JSON.stringify(event, null, 2));

  try {
    const {Contents: results} = await s3.listObjectsV2({
        Bucket: bucketName
    }).promise()

    const photos = await Promise.all(results!
      .filter((result: S3.Object) => result.Key?.includes('-20_'))
      .map((result: S3.Object) => getMetadata(result)))

    console.log('photos', JSON.stringify(photos, null, 2))

    return {
        statusCode: 200,
        headers: {
          "Access-Control-Allow-Headers" : 'Content-Type',
          "Access-Control-Allow-Origin": event.headers.origin || 'https://memerson.net',
          "Access-Control-Allow-Methods": "OPTIONS,POST,GET"
        },
        body: JSON.stringify(photos)
    }
  } catch (e: any) {
      return {
          statusCode: 500,
          body: e.message,
      }
  }
}

File: infrastructure/lambda/list_photos.py
import json
import logging
import re
from collections import defaultdict

import boto3

logger = logging.getLogger()
logger.setLevel(logging.INFO)

s3 = boto3.client('s3')
BUCKET = 'memerson-public-photos'


def list_photos_handler(event, context):
    logger.info(f'Event: {event}')
    headers = {
        "Access-Control-Allow-Headers": 'Content-Type',
        "Access-Control-Allow-Origin": event['headers'].get(
            'origin') or 'https://memerson.net, http://localhost:3000',
        "Access-Control-Allow-Methods": "OPTIONS,POST,GET"
    }

    try:
        photos = list_photos()
        response = {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps(photos)
        }
        logger.info(f'Response: {response}')
        return response
    except Exception as e:
        logger.error(f'Error listing photos: {e}')
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'message': e})
        }


def s3_url(bucket, *args):
    return f"https://{bucket}.s3.amazonaws.com/{'/'.join(args)}"


def get_dimensions(url):
    r = re.search(r'([\d]+)x([\d]+)', url)
    if not r:
        return 0, 0
    width = int(r.group(1))
    height = int(r.group(2))
    return width, height


def list_photos():
    s3_response = s3.list_objects_v2(Bucket=BUCKET)
    contents = s3_response.get('Contents', [])
    photo_keys = defaultdict(list)
    for k, v in (obj.get('Key').rsplit('/', 1) for obj in contents if obj.get('Key')):
        photo_keys[k].append(v)
    logger.info(f'Photo keys: {photo_keys}')
    photos = []
    for k, v in photo_keys.items():
        v.sort(key=lambda i: get_dimensions(i)[0])
        width, height = get_dimensions(v[0])
        photos.append({
            'src': s3_url(BUCKET, k, v[0]),
            'srcSet': [s3_url(BUCKET, k, f'{photo} {i + 1}x') for i, photo in enumerate(v)],
            # 'sizes': [],
            'width': width,
            'height': height,
        })
    return photos
